core/src/loader.rs

```diff
@@ -2107,4 +2217,449 @@ impl<'gc> Loader<'gc> {
             }
         }
     }
+
+    /// Loader to process callbacks for a file selection dialog
+    pub fn file_dialog_loader(
+        &mut self,
+        player: Weak<Mutex<Player>>,
+        dialog: DialogResultFuture,
+    ) -> OwnedFuture<(), Error> {
+        let handle = match self {
+            Loader::FileDialog { self_handle, .. } => {
+                self_handle.expect("Loader not self-introduced")
+            }
+            _ => return Box::pin(async { Err(Error::NotFileDialogLoader) }),
+        };
+
+        let player = player
+            .upgrade()
+            .expect("Could not upgrade weak reference to player");
+
+        Box::pin(async move {
+            let dialog_result = dialog.await;
+
+            // Dialog is done, allow opening new dialogs
+            player.lock().unwrap().ui_mut().close_file_dialog();
+
+            // Fire the load handler.
+            player.lock().unwrap().update(|uc| -> Result<(), Error> {
+                let loader = uc.load_manager.get_loader(handle);
+                let target_object = match loader {
+                    Some(&Loader::FileDialog { target_object, .. }) => target_object,
+                    None => return Err(Error::Cancelled),
+                    _ => return Err(Error::NotFileDialogLoader),
+                };
+
+                println!("File dialog handle 1");
+
+                let file_ref = match target_object.native() {
+                    NativeObject::FileReference(fr) => fr,
+                    _ => panic!("Shouldn't happen"),
```

Comment:
More descriptive panic message would be nice- maybe "NativeObject must be FileReference"?